File: testapps/MPClientObjCFrm/MPClientObjCFrm/ViewController.m
//
//  ViewController.m
//  MPClientObjCFrm
//
//  Created by Richard Rosiak on 2/5/15.
//  Copyright (c) 2015 Richard Rosiak. All rights reserved.
//

#import "ViewController.h"
#import <GRIDMobileSDK/GRIDMobileSDK.h>

@interface ViewController ()

@property (nonatomic, strong) PGMClient *gridClient;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



- (IBAction)signInTapped:(id)sender {
    self.Messages.text = @"";
    
    PGMAuthOptions *clientOptions = [[PGMAuthOptions alloc] initWithClientId:@"wkLZmUJAsTSMbVEI9Po6hNwgJJBGsgi5"
                                                             andClientSecret:@"SAftAexlgpeSTZ7n"
                                                              andRedirectUrl:@"http://int-piapi.stg-openclass.com/pi_group12client"];
    
    self.gridClient = [[PGMClient alloc] initWithEnvironmentType:PGMStagingEnv andOptions:clientOptions];
    
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(signInError:) name:@"LoginFailure" object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(signInComplete:) name:@"LoginSuccess" object:nil];
    
    //__block PGMAuthenticatedContext *authContextToStore = nil;
    //__block NSInteger interval;
    
    AuthenticationRequestComplete onComplete = ^(PGMAuthResponse *response) {
        //dispatch_async(dispatch_get_main_queue(), ^{ // execute the completion block on the main thread, if necessary
            if (response.error) {
                //error condition during login
                //self.Messages.text = response.error.description;
                dispatch_async(dispatch_get_main_queue(), ^{
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"LoginFailure" object:response];
                });
            }
            else {
                //login success - authenticated context returned
                //authContextToStore = response.authContext;
                //interval = response.authContext.tokenExpiresIn;
                NSLog(@"On complete success!! Access token is %@", response.authContext.accessToken);
                dispatch_async(dispatch_get_main_queue(), ^{
                    [[NSNotificationCenter defaultCenter] postNotificationName:@"LoginSuccess" object:response];
                });
            }
        //});
    };
    
    [[self.gridClient authenticator] authenticateWithUserName:self.Username.text
                                                  andPassword:self.password.text
                                                   onComplete:onComplete];
    
    //[NSThread sleepForTimeInterval:2.0];
    //NSLog(@"Done sleeping - will populate msg view...and access token from auth context to store is: %@", authContextToStore. accessToken);
    //NSLog(@"Interval is: %ld", interval);
    
    //self.Messages.text = [NSString stringWithFormat:@"%@\n%@\n%ld", authContextToStore.accessToken,
    //                      authContextToStore.refreshToken, (long)authContextToStore.tokenExpiresIn];
}

-(void) signInComplete:(NSNotification*) notification {
    
    [self removeObservers];
    
    NSLog(@"Sign in complete...");
    PGMAuthResponse *respnseFromNotification = (PGMAuthResponse*)notification.object;
    
    self.Messages.text = [NSString stringWithFormat:@"%@\n%@\n%ld", respnseFromNotification.authContext.accessToken,
                                                respnseFromNotification.authContext.refreshToken,
                                                (long)respnseFromNotification.authContext.tokenExpiresIn];
}

-(void) signInError:(NSNotification*) notification {
    
    [self removeObservers];
    
    NSLog(@"Failure trying to sign in...");
    PGMAuthResponse *respnseFromNotification = (PGMAuthResponse*)notification.object;
    self.Messages.text = respnseFromNotification.error.description;
}

- (void)removeObservers {
    
    [[NSNotificationCenter defaultCenter] removeObserver:self name:@"LoginFailure" object:nil];
    [[NSNotificationCenter defaultCenter] removeObserver:self name:@"LoginSuccess" object:nil];
}

@end

File: testapps/MPClientSwiftFrm/MPClientSwiftFrm/ViewController.swift
//
//  ViewController.swift
//  MPClientSwiftFrm
//
//  Created by Richard Rosiak on 2/6/15.
//  Copyright (c) 2015 Richard Rosiak. All rights reserved.
//

import UIKit
import GRIDMobileSDK

class ViewController: UIViewController {
    
    var client: PGMClient!
    
    @IBOutlet var username: UITextField!
    @IBOutlet var password: UITextField!
    @IBOutlet var messages: UITextView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        
        var options = PGMAuthOptions(clientId: "wkLZmUJAsTSMbVEI9Po6hNwgJJBGsgi5",
            andClientSecret: "SAftAexlgpeSTZ7n", andRedirectUrl: "http://int-piapi.stg-openclass.com/pi_group12client")
        
        self.client = PGMClient(environmentType: .StagingEnv, andOptions: options)
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    @IBAction func signInTapped(sender: AnyObject) {
        self.performSignInWith(self.username.text, password: self.password.text)
    }
    

    func performSignInWith(username: String, password: String) {
        NSOperationQueue().addOperationWithBlock({
            self.client!.authenticator.authenticateWithUserName(username, andPassword: password, onComplete: { (authResponse: PGMAuthResponse!) -> Void in
                NSOperationQueue.mainQueue().addOperationWithBlock({
                    self.signInComplete(authResponse)
                })
            })
        })
    }
    
    func signInComplete(response: PGMAuthResponse) {
        self.messages.text = ""
        
        println("In signIn complete...")
        if (response.error != nil) {
            println("In signIn complete - error \(response.error.description)")
            
            self.messages.text = response.error.description
            
        } else {
            println("Login on complete - success!!")
            println("The access token is: \(response.authContext.accessToken)")
            self.messages.text = "\(response.authContext.accessToken)\n\(response.authContext.refreshToken)\n\(response.authContext.tokenExpiresIn)\n\(response.authContext.userIdentityId)\n\(response.authContext.username)"
        }
    }
}

